### PR TITLE
feat(backend): record custom user tokens

### PR DESCRIPTION
**Pull request by roman-kashitsyn**

### CODE REVIEW

File: src/backend/src/lib.rs [hunk 1]
```diff
@@ -1,33 +1,143 @@
...
+    pub contract_address: String,
+    pub chain_id: ChainId,
+    pub symbol: Option<String>,
+    pub decimals: Option<u8>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+struct StoredPrincipal(Principal);
+
+impl Storable for StoredPrincipal {
```
Comment (peterpeterparker) [inline src/backend/src/lib.rs:117]: Out of curiosity, doesn't unbounded tree map handle automatically principal now?
i.e. this implementation `StoredPrincipal` is a must?

File: src/backend/src/lib.rs [hunk 2]
```diff
@@ -1,33 +1,143 @@
...
+    const BOUND: Bound = Bound::Bounded {
+        max_size: 29,
```
Comment (peterpeterparker) [inline src/backend/src/lib.rs:119]: it's not 32 bytes for a Principal?

Comment (bitdivine) [reply src/backend/src/lib.rs:119]: I know that account identifiers are typically 32 bytes long.  This test passes:
```
/// We assume that new account identifiers are 32 bytes long.
///
/// Note: Old-style account identifiers were 28 bytes long.
#[test]
fn account_identifier_has_32_bytes() {
    let account_identifier = icp_ledger::AccountIdentifier::from(ic_base_types::PrincipalId::new_user_test_id(99));
    assert_eq!(
        account_identifier.to_vec().len(),
        AccountStorageKey::ACCOUNT_IDENTIFIER_MAX_BYTES // Note: This is 32.
    );
}
```

Comment (bitdivine) [reply src/backend/src/lib.rs:119]: I would check the spec and the code that defines principal to be sure though.

Comment (bitdivine) [reply src/backend/src/lib.rs:119]: The [candid code that defines Principal](https://docs.rs/candid/0.9.10/src/candid/types/principal.rs.html#85) says 29 bytes and mentions a 4 byte CRC.

Comment (bitdivine) [reply src/backend/src/lib.rs:119]: By the way, I just got a notification from GitHub saying that a review was requested.  I guess that given that this is OISY wallet, the review is probably wanted from @peterpeterparker but I am happy to pitch in as well, if wanted.  I don't know the business context but can read Rust, if wanted.  Please let me know.

File: src/backend/src/lib.rs (hunk 1 above)
Comment (roman-kashitsyn) [reply src/backend/src/lib.rs:117]: Principals aren't storable, unbounded keys/values are orthogonal to the encoding.

File: src/backend/src/lib.rs (hunk 2 above)
Comment (peterpeterparker) [reply src/backend/src/lib.rs]: Yeah nvm my comment, my brain was off for a minute.

File: src/backend/src/lib.rs (hunk 1 above)
Comment (peterpeterparker) [reply src/backend/src/lib.rs:117]: Weird. [Here](https://github.com/junobuild/juno/blob/1f4043de026e0e04d6731f3c91ce550d737a0c41/src/satellite/src/db/impls.rs#L19) I implement storable for struct that contains a principal and I do not get any errors, at least at build time. I should double check then at runtime.

Comment (peterpeterparker) [reply src/backend/src/lib.rs:117]: Thanks for the feedback.

Comment (peterpeterparker) [reply src/backend/src/lib.rs:117]: Ah it's because the principal is used within the value, not the key!
Sorry for the noise.

**Pull request merged**